### ISSUE TITLE
Undefined reference of `operator<<(ostream&, const string&)` when compiling with clang 10 and libc++ 10 on Ubuntu 16.04.6 LTS

### ISSUE THREAD

**Issue by yeputons** [doctest/doctest#356]
## Description

I tried to compile a minimal Doctest example of two files on Ubuntu 16.04.6 LTS with clang 10 and libc++ 10, which resulted in link errors.

I assume it's because `<iosfwd>` is included without `<iostream>`, and `some_ostream << some_string;` compiles, while yielding wrong ABI,

### Steps to reproduce

Create two files:
```
// a.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
```
and
```
// b.cpp
#include <string>
#include "doctest.h"

void foo() {
    std::string a, b;
    CHECK(a == b);
}
```

and run `clang++-10 -O2 -stdlib=libc++ -std=c++17 a.cpp b.cpp`. I expected the example to compile, but it resulted in

```
/tmp/b-c494e1.o: In function `doctest::String doctest::detail::stringifyBinaryExpr<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)':
b.cpp:(.text._ZN7doctest6detail19stringifyBinaryExprINSt3__112basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEES8_EENS_6StringERKT_PKcRKT0_[_ZN7doctest6detail19stringifyBinaryExprINSt3__112basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEES8_EENS_6StringERKT_PKcRKT0_]+0x23): undefined reference to `std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<< <char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
b.cpp:(.text._ZN7doctest6detail19stringifyBinaryExprINSt3__112basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEES8_EENS_6StringERKT_PKcRKT0_[_ZN7doctest6detail19stringifyBinaryExprINSt3__112basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEES8_EENS_6StringERKT_PKcRKT0_]+0x5e): undefined reference to `std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<< <char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Extra information

If I add `#include <iostream>` on top of `b.cpp`, it compiles successfully.

Strictly speaking, `<iosfwd>` [does not define](https://en.cppreference.com/w/cpp/header/iosfwd) any operations on `ostream`, but `<string>` [does](https://en.cppreference.com/w/cpp/header/string). So I'm not sure if this is an issue with doctest not including `<iostream>` or libc++ including `<string>` without all necessary dependencies.

* doctest version: **v2.3.7**
* Operating System: **Ubuntu 16.04.6 LTS**
* Compiler+version: `clang version 10.0.0-++20200412072846+50d7e5d5e7d-1~exp1~20200412053445.132`

I've used clang and libc++ from [apt.llvm.org](https://apt.llvm.org/) (`sudo apt install clang-10 libc++-10-dev libc++abi-10-dev libc++1-10 libc++abi1-10`), and [doctest 2.3.7](https://raw.githubusercontent.com/onqtam/doctest/2.3.7/doctest/doctest.h):
```
clang version 10.0.0-++20200412072846+50d7e5d5e7d-1~exp1~20200412053445.132 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by onqtam** [comment]
Thanks for reporting! Well this sounds surprisingly similar as this issue: https://github.com/onqtam/doctest/issues/126

This happens only with clang and libc++ (in contrast to libstdc++). Personally I think this is a toolchain issue as it's weird for me for the right overload to be found and for it to compile successfully, but not to link. However, I might have to finally do something about this... But not sure when I'll get to that. I'd advise you to just include `<stream>` in `b.cpp` for now.

**Comment by yeputons** [comment]
Looks like it was already reported to libc++ almost a year ago: [Bug 41563 - `operator<<(ostream, string)` declared in `<string>` but defined in `<ostream>`](https://bugs.llvm.org/show_bug.cgi?id=41563). I've bumped the thread.

**Comment by h5rdly** [comment]
I've an issue with Clang 10 as well on Windows. Not sure if it's a separate ticket though.

For the tutorial code in `test.cpp`:

```
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

int factorial(int number) { return number <= 1 ? number : factorial(number - 1) * number; }

TEST_CASE("testing the factorial function") {
    CHECK(factorial(1) == 1);
    CHECK(factorial(2) == 2);
    CHECK(factorial(3) == 6);
    CHECK(factorial(10) == 3628800);
}
```
With `clang++ -o test.exe test.cpp`, I get:

```
# clang++ -o test.exe test.cpp
test-94ae87.o : error LNK2019: unresolved external symbol fileno referenced in function "void 
__cdecl doctest::`anonymous namespace'::color_to_stream(class std::basic_ostream<char,struct 
std::char_traits<char> > &,enum A0xEEF7E3AD::Color::Enum)" (?color_to_stream@?
A0xEEF7E3AD@doctest@@YAXAEAV?$basic_ostream@DU?
$char_traits@D@std@@@std@@W4Enum@Color@1@@Z) test-94ae87.o : error LNK2019: 
unresolved external symbol isatty referenced in function "void __cdecl doctest::`anonymous 
namespace'::color_to_stream(class std::basic_ostream<char,struct std::char_traits<char> > &,enum 
A0xEEF7E3AD::Color::Enum)" (?color_to_stream@?A0xEEF7E3AD@doctest@@YAXAEAV?
$basic_ostream@DU?$char_traits@D@std@@@std@@W4Enum@Color@1@@Z) 
test.exe : fatal error LNK1120: 2 unresolved externals
clang++: error: linker command failed with exit code 1120 (use -v to see invocation)
```

I'm on Windows 10 64 bit, my `clang --version` details:
```
clang version 10.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
```

**Comment by onqtam** [comment]
@h5rdly I think that's a separate issue - it cannot find [`isatty`](https://github.com/onqtam/doctest/blob/2.3.7/doctest/doctest.h#L3757) and I'm clueless as to why...

**Comment by h5rdly** [comment]
Don't know if it helps, but if run the same command with `-fuse-ld=lld` I get:

```
lld-link: error: undefined symbol: fileno
>>> referenced by C:\Users\bobo\AppData\Local\Temp\test-57d48b.o:(void __cdecl doctest::`anonymous 
namespace'::color_to_stream(class std::basic_ostream<char, struct std::char_traits<char>> &, 
enum 0xEEF7E3AD::Color::Enum))

lld-link: error: undefined symbol: isatty
>>> referenced by C:\Users\bobo\AppData\Local\Temp\test-57d48b.o:(void __cdecl doctest::`anonymous
 namespace'::color_to_stream(class std::basic_ostream<char, struct std::char_traits<char>> &, 
enum 0xEEF7E3AD::Color::Enum))
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

How do you start debugging something like this?

Clang 10 has std::span, and concepts, and seems to be fully supported by MSVC 16.6 - 
https://github.com/microsoft/STL/issues/359

It would be a shame to not be able to use doctest with it :/

**Comment by onqtam** [comment]
@h5rdly well you could disable the console colors for now by defining [`DOCTEST_CONFIG_COLORS_NONE`](https://github.com/onqtam/doctest/blob/master/doc/markdown/configuration.md#doctest_config_colors_none) globally. I think this should be spun as a separate issue - I bet there's a trivial fix but I just don't have the setup to debug & experiment locally...

**Comment by h5rdly** [comment]
Works!

Do you want me to open a separate ticket then?

**Comment by onqtam** [comment]
@h5rdly sure :)

**Comment by yeputons** [comment]
Here is another workaround: disable output prettification (and the need in `<<`) by adding extra parens in `CHECK`, see https://github.com/cquery-project/cquery/issues/649 .

**Comment by yeputons** [comment]
Update for clarity:

1. In the examples above doctest does not include `<iosfwd>` at all because `DOCTEST_CONFIG_USE_STD_HEADERS` is not defined. It provides its own forward declarations which is not standard complaint.
2. However, libc++ is still affected by the bug reported.
3. If one defines `DOCTEST_CONFIG_USE_STD_HEADERS`, then doctest includes both `<iosfwd>` and `<ostream>` (for some reason) which successfully works around the libc++'s bug.

**Comment by nextsilicon-itay-bookstein** [comment]
I think that the forward declarations when `!DOCTEST_CONFIG_USE_STD_HEADERS` implicitly assume that `operator<<(ostream&, std::string const&)` is defined _somewhere_, which indirectly assumes that either:
1. The stdlib shared object defines it (which appears to **not** be the case for reasonably-recent libc++ builds)
2. A translation unit statically linked into the executable defines it (which means that at least one translation unit _that actually uses the function_ needs to have a visible definition, which for libc++ necessitates including <ostream>)

I know that clang assumes a definition is available for some template symbols and doesn't codegen them into the client executable (also known as LLVM's `available_externally` linkage type), but it looks like this doesn't apply to these particular ones?
Does anything in the standard make libc++'s behavior of not having a definition for these in libc++.so itself non-compliant?

**Comment by yeputons** [comment]
I've got a soft report from a student that Apple Clang is probably affected as well (it uses libc++), which is the default on macOS

**Comment by onqtam** [comment]
@yeputons just pushed a commit in dev that always uses the std headers with clang and libc++ - could you ask your student to try with this version and let me know if it works:
https://github.com/doctest/doctest/blob/b59552dc43c4e68f58f0ab4163b5d8f5f9e28837/doctest/doctest.h

**Comment by yeputons** [comment]
@onqtam Thank you, it fixes the original issue for me (at least with Clang 12 on Ubuntu 20.04.3 LTS; I was able to reproduce it with a384864346f48c7ff5608051c176e7bd505712ae and fix with b59552dc43c4e68f58f0ab4163b5d8f5f9e28837) and my student (Apple Clang).

**Comment by onqtam** [comment]
closing this - will release an official version soon!
